Adding types for are.na / arena-js

huntercaron opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

----

# Adding Types for Are.na API javascript wrapper
Adds types to the `are.na` package on npm for the wrapper functions themselves as well as the responses. 

Included are generic are.na types for `Channel, Block, User, ...` that can also be imported individually to type your `.ts` files if you'd like (and are used to type the responses).

Comments were pulled from [the package repo](https://github.com/ivangreene/arena-js) and from the [Are.na API docs](https://dev.are.na/documentation/channels). Some were slightly edited to work in context. 

## Why?
Whereas most people don't need to worry that these types exist as I know Typescript isn't for everyone, I hope that for people who do use typescript can have a more enjoyable experience making things with the are.na API with this. These are the two ways I think about it:

**Inline auto-complete & information**
When writing code, you don't have to remember the shape of the blocks response as it appears in-line as you type.

**General author confidence**
When building things that other people are going to use, I generally find I can be more confident about having less bugs, or easily finding them, when using types.

huntercaron commented:
@ivangreene if you have any thoughts on this as it's your library. I tried to follow the docs as closely as possible

ivangreene commented:
@huntercaron I think this looks good overall, though I'm by no means a typescript expert and I also haven't worked on the project in a while. @broskoski might want to review too, he has taken over maintenance on the project

huntercaron commented:
thanks @ivangreene - @broskoski let me know if you have any questions

martin-badin commented on types/are.na/index.d.ts:
```diff
@@ -0,0 +1,621 @@
+     * @param accessToken - Your are.na API access token
+     * @param baseURL - Base URL to make requests on (default: https://api.are.na/v2/)
+     */
+    constructor(config?: { accessToken?: string; baseURL?: string });
```
missing `authToken` and `requestHandler` 
https://github.com/ivangreene/arena-js/blob/master/index.js#L32
https://github.com/ivangreene/arena-js/blob/master/index.js#L40

martin-badin commented on types/are.na/index.d.ts:
```diff
@@ -0,0 +1,621 @@
+     * @param baseURL - Base URL to make requests on (default: https://api.are.na/v2/)
+     */
+    constructor(config?: { accessToken?: string; baseURL?: string });
+
```
Missing `group`

https://github.com/ivangreene/arena-js/blob/master/index.js#L58

martin-badin commented on types/are.na/index.d.ts:
```diff
@@ -0,0 +1,621 @@
+        /**
+         * Creates a new channel. Called as channel(title).create(status). Title is required, status is optional.
+         */
+        create(status: Arena.ChannelStatus): Promise<void>;
```
it seems, in this function is a missing `title` by your documentation but then it is the same as https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51403/files#diff-91c4eda792aace8177f766a6c99725c79a920925a768795e12bd0f3bfdd80ca0R86 so I think that it is unnecessary

huntercaron replied:
This function is overloaded with either `(status)` or `(title, status)` as documented inline here https://github.com/ivangreene/arena-js/blob/master/index.js#L103-L104

martin-badin commented on types/are.na/index.d.ts:
```diff
@@ -0,0 +1,621 @@
+        /**
+         * Remove a block from the channel
+         */
+        deleteBlock(blockID: number): Promise<void>;
```
`blockID` is string
https://github.com/ivangreene/arena-js/blob/master/index.js#L132

martin-badin commented on types/are.na/index.d.ts:
```diff
@@ -0,0 +1,621 @@
+     * @param params - Additional parameters
+     */
+    user(
+        id: number,
```
`id` is string
https://github.com/ivangreene/arena-js/blob/master/index.js#L153

huntercaron replied:
Good catch - I have tested and these functions will work with either string or number and it makes more ergonomic sense as a number. I've changed this and the `id` cases above to support `string | number` to be more accurate

huntercaron commented:
thanks for the thoughtful feedback @martin-badin  - great catches. I have addressed all the feedback in the past few commits.

The group API isn't documented anywhere yet as far as I can tell so I've gone ahead and documented it as well as I could with the current response.

martin-badin commented:
> @huntercaron I'm noticing when I'm installing in 2024 I'm getting a few peer dependency issues from eslint and "Cannot read properties of null (reading 'matches')"
> 
> ```
> npm WARN ERESOLVE overriding peer dependency
> npm WARN ERESOLVE overriding peer dependency
> npm WARN ERESOLVE overriding peer dependency
> npm WARN ERESOLVE overriding peer dependency
> npm WARN While resolving: @eslint/eslintrc@2.1.4
> npm WARN Found: peer eslint-plugin-jsdoc@">=22.1.0" from eslint-config-eslint@7.0.0
> npm WARN node_modules/.pnpm/@eslint+eslintrc@2.1.4/node_modules/@eslint/eslintrc/node_modules/eslint-config-eslint
> npm WARN   dev eslint-config-eslint@"^7.0.0" from @eslint/eslintrc@2.1.4
> npm WARN   node_modules/.pnpm/@eslint+eslintrc@2.1.4/node_modules/@eslint/eslintrc
> npm WARN 
> npm WARN Could not resolve dependency:
> npm WARN peer eslint-plugin-jsdoc@">=22.1.0" from eslint-config-eslint@7.0.0
> npm WARN node_modules/.pnpm/@eslint+eslintrc@2.1.4/node_modules/@eslint/eslintrc/node_modules/eslint-config-eslint
> npm WARN   dev eslint-config-eslint@"^7.0.0" from @eslint/eslintrc@2.1.4
> npm WARN   node_modules/.pnpm/@eslint+eslintrc@2.1.4/node_modules/@eslint/eslintrc
> npm WARN ERESOLVE overriding peer dependency
> npm WARN ERESOLVE overriding peer dependency
> npm WARN ERESOLVE overriding peer dependency
> npm WARN ERESOLVE overriding peer dependency
> npm ERR! Cannot read properties of null (reading 'matches')
> ```

You have an issue with the javascript code, Not with this package. This package is getting just types to the javascript module.
